### ISSUE TITLE
For bundling, desktop apps should canonicalize assets automatically.

### ISSUE THREAD

**Issue by jkelleyrtp** [DioxusLabs/dioxus#232]
## Specific Demand

When resolving assets, we should determine of the current program is contained within a bundle. If it is, then we should cananocalize the asset differently to be relative to the `.app`/`.deb`/`.rpm`/`.exe` for a smoother distribution experience.

- [x] macOS

These targets do not provide a straightforward way of bundling assets. I've opted to make it the asset dir configurable instead.
- [ ] windows
- [ ] linux (rpm)
- [ ] linux (deb)

## Implement Suggestion

https://github.com/burtonageo/cargo-bundle/pull/93/files

For macos: 
```rust

        // TODO: support for other platforms
         #[cfg(target_os = "macos")]
         {
             let bundle = core_foundation::bundle::CFBundle::main_bundle();
             let bundle_path = bundle.path()?;
             let resources_path = bundle.resources_path()?;
             let absolute_resources_root = bundle_path.join(resources_path);
             let canonical_resources_root = dunce::canonicalize(absolute_resources_root).ok()?;
             Some(canonical_resources_root)
         }
```

I have yet to figure out linux or windows ways of doing this.

**Comment by autarch** [comment]
At least on Linux, we shouldn't expect assets like images to share a root with the executable. That may be the case in some packaging scenarios, but a typical apt package will have images somewhere like `/usr/share/myapp` and the application itself at `/usr/bin/myapp`.

If the goal is to support single-file distributions, then there are probably ways to compile the assets into the executable itself. Maybe Dioxus could offer support for accessing those assets for desktop apps.

**Comment by jkelleyrtp** [comment]
Is there a way to get the directory for Linux assets? So installed packages know where they're being launched from? Im sure a crate like bevy has figured this out.

**Comment by autarch** [comment]
For regular packages, there are things like the [Debian Filesystem Hierarchy Standard](https://www.debian.org/doc/packaging-manuals/fhs/fhs-3.0.html). There's [one for Fedora](https://docs.fedoraproject.org/en-US/Fedora/14/html/Storage_Administration_Guide/s1-filesystem-fhs.html), too.

I looked for crates that might help with this but I only found crates that abstract per-user directories across operating systems. I didn't find any that do the same for system directories.

I dug into Bevy a bit to see how it handles it, but I think it does some sort of bundling for distribution, so that you don't have to deal with the FHS at all. Nothing in the docs talks about handling things like the FHS.

**Comment by jkelleyrtp** [comment]
We should optimize for the case of cargo-bundle - and I'm not entirely sure how their resources system works.

**Comment by jkelleyrtp** [comment]
Okay so cargo bundle is broken on windows.... Sounds like we have to fix cargo bundle first. Seems like people use cargo-wix for windows. Still not sure what it is for linux.

#278 

If we ever get some money to put towards hiring people, then maybe we should focus on resolving assets. Or somehow merge our desktop efforts into Tauri.

**Comment by cxgreat2014** [comment]
Maybe here can help this issue: https://github.com/tauri-apps/tauri/tree/4c84559e1f3019e7aa2666b10a1a0bd97bb09d24/tooling/bundler/src/bundle

**Comment by Demonthos** [comment]
https://github.com/DioxusLabs/cli/pull/118 should fix this issue on windows. It uses Tauri's bundler. We need to test this on MacOS, and Linux

**Comment by jrouaix** [comment]
Well, it does not seems to work for linux right now, for example :

```rust
fn main() -> anyhow::Result<()> {
  // Dioxus setup
  let config = dioxus_desktop::Config::new()
    .with_custom_head(r#"<link rel="stylesheet" href="public/main.css">"#.to_string());

  dioxus_desktop::launch_cfg(app::App, config);
  
  Ok(())
}
```

when I launch the app with `dx serve --hot-reload --platform desktop` the style is applied, no problem.

Then I bundle the app with `dx bundle --release` 
the css file will be install in some `/usr/lib/...` fodler.

Then the css is not found anymore (same for .dev and .AppImage)

**Comment by ealmloff** [comment]
You can check the CARGO environment variable to see if the program is currently being run by cargo. We should integrate that into this code: https://github.com/DioxusLabs/dioxus/blob/master/packages/desktop/src/protocol.rs#L122 to get the asset root

**Comment by jkelleyrtp** [comment]
This is fixed now, but we still need to spec out exactly how android will load assets in prod

Fixed byt #2779

**Comment by phkoester** [comment]
How is this fixed? If I create an AppImage on Linux with dx 0.6.2, the assets are not found if the app is run from the image. They reside in `squashfs-root/usr/lib/HotDog/assets`.

What's even worse, if I debug-output the assets, my private workspace is revealed:

```
MAIN_CSS: Asset { bundled: BundledAsset { absolute_source_path: "/home/phkoester/project/hot_dog/assets/styling/main.css", bundled_path: "main-840b34480908b3b9.css", options: Unknown }, keep_link_section: 0x55e3cf006650 }
```

That's an information I would not expect to be included in the final distributed binary.
